pkg/kotsadmsnapshot/backup.go

```diff
@@ -191,254 +201,512 @@ func CreateApplicationBackup(ctx context.Context, a *apptypes.App, isScheduled b
+	}
+
+	return metadata, nil
+}
+
+// getECInstanceBackupMetadata returns metadata about the embedded cluster for use in creating an
+// instance backup.
+func getECInstanceBackupMetadata(ctx context.Context, ctrlClient ctrlclient.Client) (*ecInstanceBackupMetadata, error) {
+	if !util.IsEmbeddedCluster() {
+		return nil, nil
+	}
+
+	installation, err := embeddedcluster.GetCurrentInstallation(ctx, ctrlClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current installation: %w", err)
+	}
+
+	seaweedFSS3ServiceIP, err := embeddedcluster.GetSeaweedFSS3ServiceIP(ctx, ctrlClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get seaweedfs s3 service ip: %w", err)
+	}
+
+	return &ecInstanceBackupMetadata{
+		installation:         *installation,
+		seaweedFSS3ServiceIP: seaweedFSS3ServiceIP,
+	}, nil
+}
+
+// getInfrastructureInstanceBackupSpec returns the velero backup spec for the instance backup. This
+// is either the kotsadm backup or the combined backup if this is not using improved DR.
+func getInfrastructureInstanceBackupSpec(ctx context.Context, k8sClient kubernetes.Interface, metadata instanceBackupMetadata, hasAppSpec bool) (*velerov1.Backup, error) {
+	// veleroBackup is the kotsadm backup or combined backup if usesImprovedDR is false
+	veleroBackup := &velerov1.Backup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:         "",
+			GenerateName: "instance-",
+			Annotations:  map[string]string{},
+		},
+		Spec: velerov1.BackupSpec{
+			StorageLocation:         "default",
+			IncludedNamespaces:      []string{metadata.kotsadmNamespace},
+			ExcludedNamespaces:      []string{},
+			IncludeClusterResources: ptr.To(true),
+			OrLabelSelectors:        instanceBackupLabelSelectors(metadata.ec != nil),
+			OrderedResources:        map[string]string{},
+			Hooks: velerov1.BackupHooks{
+				Resources: []velerov1.BackupResourceHookSpec{},
+			},
+		},
+	}
+
+	if util.AppNamespace() != metadata.kotsadmNamespace {
+		veleroBackup.Spec.IncludedNamespaces = append(veleroBackup.Spec.IncludedNamespaces, util.AppNamespace())
+	}
+
+	isKurl, err := kurl.IsKurl(k8sClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to check if cluster is kurl")
+	}
+	if isKurl {
+		veleroBackup.Spec.IncludedNamespaces = append(veleroBackup.Spec.IncludedNamespaces, "kurl")
+	}
+
+	isKotsadmClusterScoped := k8sutil.IsKotsadmClusterScoped(ctx, k8sClient, metadata.kotsadmNamespace)
+	if !isKotsadmClusterScoped {
+		// in minimal rbac, a kotsadm role and rolebinding will exist in the velero namespace to give kotsadm access to velero.
+		// we backup and restore those so that restoring to a new cluster won't require that the user provide those permissions again.
+		veleroBackup.Spec.IncludedNamespaces = append(veleroBackup.Spec.IncludedNamespaces, metadata.backupStorageLocationNamespace)
+	}
+
+	for _, appMeta := range metadata.apps {
+		// Don't merge the backup spec if we are using the new improved DR.
+		if !hasAppSpec {
```

Comment:
app spec could mean many different things in kots. maybe rename to `hasAppBackup`?

Reply:
because the function right below this line is called `mergeAppBackupSpec`, which took me a bit to understand.

Reply:
done